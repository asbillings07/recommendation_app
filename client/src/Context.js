import React, { Component } from 'react';
import Data from './Data';
import Cookies from 'js-cookie';
const Context = React.createContext();

/**
 * Class that Provides context
 * SignIn and SignOut Methods
 *
 */

export class Provider extends Component {
  data = new Data();

  state = {
    authorizedUser: Cookies.getJSON('authorizedUser') || null,
    token: Cookies.getJSON('token') || null,
  };

  render() {
    const { authorizedUser } = this.state;

    const value = {
      authorizedUser,
      data: this.data,
      actions: {
        signIn: this.signIn,
        signOut: this.signOut,
      },
    };

    return (
      <Context.Provider value={value}>{this.props.children}</Context.Provider>
    );
  }

  /** SignIn Method - Signs in user and sets authorized user in cookies */

  signIn = async (email, password) => {
    const creds = { email, password };
    const user = await this.data.login(creds);
    if (user) {
      console.log(user);
      console.log(user.token);
      this.setState(() => {
        return {
          authorizedUser: user.user,
          token: user.token,
        };
      });
      Cookies.set('authorizedUser', JSON.stringify(user.user), { expires: 1 });
      Cookies.set('token', JSON.stringify(user.token), { expires: 1 });
    }
    return user;
  };

  /**SignOut Method - Signs out user and removes cookies */

  signOut = () => {
    this.setState({ authorizedUser: null, token: null });
    Cookies.remove('authorizedUser');
    Cookies.remove('token');
    console.log('SignOut Successful');
  };
}

export const Consumer = Context.Consumer;

/**
 * A higher-order component that wraps the provided component in a Context Consumer component.
 * @param {class} Component - A React component.
 * @returns {function} A higher-order component.
 */

export default function withContext(Component) {
  return function ContextComponent(props) {
    return (
      <Context.Consumer>
        {context => <Component {...props} context={context} />}
      </Context.Consumer>
    );
  };
}
