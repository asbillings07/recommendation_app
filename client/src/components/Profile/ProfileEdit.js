import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import Forms from '../Forms';
import axios from 'axios';
import Config from '../../Config';
import { Form, Container, Row, Col } from 'react-bootstrap';
import { notify } from 'react-notify-toast';

const ProfileEdit = ({ context, history }) => {
  const [errors, setErrors] = useState([]);
  const [confirmed] = useState(true);
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [email, setEmail] = useState('');

  // fetch data from database with JWT to get user info
  useEffect(() => {
    const fetchData = async () => {
      const data = await axios.get(`${Config.apiBaseUrl}/users`, {
        headers: { Authorization: 'bearer ' + context.token },
      });

      if (data) {
        setFirstName(data.data.firstName);
        setLastName(data.data.lastName);
        setEmail(data.data.email);
      }
    };

    fetchData();
  }, [context.token]);

  // submits update user info to database
  const submit = () => {
    const profileInfo = {
      firstName,
      lastName,
      email,
    };
    context.data
      .updateUser(context.token, profileInfo)
      .then(errors => {
        if (errors.length) {
          setErrors([errors[0]]);
        } else {
          notify.show('Profile Updated!', 'success', 10000);
          history.push('/profile');
        }
      })
      .catch(error => console.log(error));
  };
  const cancel = () => {
    history.push(`/profile`);
  };

  return (
    <Container className="mt-3">
      <Row className="justify-content-md-center">
        <Col xs md lg="auto">
          <>
            <h1 className="mb-4">Update Your Profile</h1>
            <Forms
              cancel={cancel}
              errors={errors}
              submit={submit}
              passwordErrors={confirmed}
              submitButtonText="Update Profile"
              elements={() => (
                <React.Fragment>
                  <Form.Group>
                    <Form.Control
                      type="text"
                      name="firstName"
                      placeholder=""
                      value={firstName}
                      onChange={e => setFirstName(e.target.value)}
                    />
                  </Form.Group>
                  <Form.Group>
                    <Form.Control
                      type="text"
                      name="lastName"
                      placeholder=""
                      value={lastName}
                      onChange={e => setLastName(e.target.value)}
                    />
                  </Form.Group>
                  <Form.Group>
                    <Form.Control
                      type="text"
                      name="email"
                      value={email}
                      placeholder=""
                      onChange={e => setEmail(e.target.value)}
                    />
                  </Form.Group>
                </React.Fragment>
              )}
            />
          </>
        </Col>
      </Row>
    </Container>
  );
};

export default ProfileEdit;
