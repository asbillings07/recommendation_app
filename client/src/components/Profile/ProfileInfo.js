import React from 'react';
import { Card, Button } from 'react-bootstrap';
import styled from 'styled-components';

export default function ProfileInfo({ user }) {
  return (
    <StyledCard aria-label="Update Profile Information">
      <Card.Body>
        <Card.Title aria-label={`The first name is ${user.firstName}`}>
          First Name: {user.firstName}
        </Card.Title>
        <Card.Title
          aria-label={`The last name is ${user.lastName}`}
          className="mb-3"
        >
          Last Name: {user.lastName}
        </Card.Title>
        <Card.Title aria-label={`The email is ${user.email}`}>
          Email Address: {user.email}
        </Card.Title>
        <ConfirmedSubtitle
          aria-label={`Your email is ${
            user.confirmed ? 'CONFIRMED' : 'NOT CONFIRMED'
          }`}
          className="mb-3"
        >
          Email Status: {user.confirmed ? 'CONFIRMED' : 'NOT CONFIRMED'}{' '}
        </ConfirmedSubtitle>
        <Button
          data-testid="profile-edit"
          variant="outline-info"
          className="mr-2"
          href="/profile/edit"
        >
          Edit Profile Info
        </Button>
        <Button
          data-testid="forgotpassword"
          variant="outline-warning"
          href="/forgotpassword"
        >
          Reset Password
        </Button>
      </Card.Body>
    </StyledCard>
  );
}

const StyledCard = styled(Card)`
  width: 18rem;
`;

const ConfirmedSubtitle = styled(Card.Subtitle)`
  ${props => (props.confirmed ? 'color: red;' : 'color: green;')}
`;
