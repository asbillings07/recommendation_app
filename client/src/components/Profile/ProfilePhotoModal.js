import React from 'react';
import { Button, Modal, Card, CardGroup } from 'react-bootstrap';
import { notify } from 'react-notify-toast';
export default function ProfileModal({
  showModal,
  setModal,
  profilePhotos,
  setPhoto,
  context,
  refresh,
}) {
  const sendAvatarToDB = photo => {
    context.data.updateUserPhoto(context.token, photo).then(error => {
      if (error.length) {
        console.log(error);
      } else {
        refresh();
        notify.show('Profile Avatar Uploaded!', 'success', 10000);
      }
    });
  };

  const handleClose = () => setModal(false);

  const showAvatars = () => {
    return profilePhotos.map((photo, i) => (
      <Card key={i}>
        <Card.Img variant="top" src={photo.photoUrl} />
        <Card.Body>
          <Card.Title>{photo.name}</Card.Title>
          <Button
            onClick={() => {
              setPhoto(photo.photoUrl);
              handleClose();
              sendAvatarToDB(photo);
            }}
            variant="primary"
          >
            Choose Avatar
          </Button>
        </Card.Body>
      </Card>
    ));
  };

  return (
    <>
      <Modal show={showModal} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Choose Your Avatar</Modal.Title>
        </Modal.Header>
        <CardGroup>{showAvatars()}</CardGroup>
        <Modal.Footer>
          <Button onClick={handleClose}>Close</Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
