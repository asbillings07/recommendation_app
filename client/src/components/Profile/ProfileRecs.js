import React, { useState, useEffect } from 'react';
import { ListGroupItem, Card, Button } from 'react-bootstrap';
import styled from 'styled-components';
import ProfileModal from './ProfileModal';

const ProfileRec = ({ recommendations, context, onRefresh }) => {
  const [showModal, setShowModal] = useState(false);
  const [recid, setRecid] = useState('');

  const profileRecs = () =>
    recommendations.map(rec => (
      <ListGroupItem key={rec.id}>
        <CardTitle aria-label={`The title of the recommendation`}>
          {rec.title}
        </CardTitle>
        <Card.Text aria-label={`The location of the recommendation`}>
          Location: {rec.location}
        </Card.Text>
        <Card.Text aria-label={`The description of the recommendation`}>
          Description: {rec.description}
        </Card.Text>
        <Button variant="info" href={`/category/${rec.id}/recs/update`}>
          Edit
        </Button>
        <Button
          onClick={() => {
            setShowModal(true);
            setRecid(rec.id);
          }}
          className="float-right"
          variant="danger"
        >
          Delete
        </Button>
      </ListGroupItem>
    ));

  return (
    <>
      {profileRecs()}
      <ProfileModal
        showModal={showModal}
        setModal={setShowModal}
        context={context}
        recId={recid}
        refresh={onRefresh}
      />
    </>
  );
};

const CardTitle = styled(Card.Title)`
  color: blue;
  font-size: 20px;
`;

export default ProfileRec;
