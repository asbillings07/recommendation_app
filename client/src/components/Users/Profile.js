import React, { Component } from 'react';
import axios from 'axios';
import Config from '../../Config';
import { Card } from 'react-bootstrap';

export class Profile extends Component {
  state = {
    user: {},
  };

  componentDidMount() {
    this.fetchData();
  }

  fetchData = async () => {
    const data = await axios.get(`${Config.apiBaseUrl}/users`, {
      headers: { Authorization: 'bearer ' + this.props.context.token },
    });

    if (data) {
      this.setState({
        user: data.data,
      });
    }
  };

  render() {
    const { user } = this.state;

    return (
      <div>
        <h1>Profile</h1>
      </div>
    );
  }
}
