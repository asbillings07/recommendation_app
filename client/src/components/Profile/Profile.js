import React, { useEffect, useState } from 'react';
import axios from 'axios';
import avatar from '../../images/imgholdr-image.png';
import Config from '../../Config';
import ProfileRecommendation from './ProfileRecs';
import UserProfileInfo from './ProfileInfo';
import ProfilePhotoModal from './ProfilePhotoModal';
import { Container, Row, Col, Card, Button } from 'react-bootstrap';
import {
  Image,
  Video,
  Transformation,
  CloudinaryContext,
} from 'cloudinary-react';
import styled from 'styled-components';

export const Profile = ({ context }) => {
  const [user, setUser] = useState({});
  const [recommendation, setRecommendation] = useState([]);
  //const [selectedFile, setSelectedFile] = useState('');
  const [photo, setPhoto] = useState('sample');
  const [showModal, setShowModal] = useState(false);

  const fetchData = async () => {
    const data = await axios.get(`${Config.apiBaseUrl}/users`, {
      headers: { Authorization: 'bearer ' + context.token },
    });

    if (data) {
      setUser(data.data);
      setRecommendation(data.data.Recommendations);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  const profilePhotos = [
    {
      photoName: 'Ninja Turtle Avatar',
      photoUrl:
        'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401759/ninja-turtle.png',
    },
    {
      photoName: 'Penguin Avatar',
      photoUrl:
        'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401755/penguin-avatar.png',
    },
    {
      photoName: 'Fox Avatar',
      photoUrl:
        'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401753/fox-avatar.png',
    },
    {
      photoName: 'Ninja Avatar',
      photoUrl:
        'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401753/ninja-avatar.png',
    },
  ];

  return (
    <CloudinaryContext cloudName="demo">
      <StyledContainer>
        <Row>
          <Col sm={4}>
            <StyledCard aria-label="profile description">
              <StyledImage
                publicId={user.imageId ? user.imageId : photo}
                width="286"
                crop="scale"
              />
              <Card.Body>
                <Card.Text>
                  This is your profile. You can reset your password, view and
                  manage all of the recommendations you have created and choose
                  a profile Avatar
                </Card.Text>

                <Button onClick={() => setShowModal(true)} variant="primary">
                  Update Profile Avatar
                </Button>
              </Card.Body>
            </StyledCard>
            <UserProfileInfo user={user} />
          </Col>

          <StyledCol sm={8}>
            <h1>Manage Recommendations</h1>
            <ProfileRecommendation
              recommendations={recommendation}
              context={context}
              onRefresh={fetchData}
            />
          </StyledCol>
        </Row>
      </StyledContainer>
      <ProfilePhotoModal
        profilePhotos={profilePhotos}
        setModal={setShowModal}
        showModal={showModal}
        setPhoto={setPhoto}
        context={context}
        refresh={fetchData}
      />
    </CloudinaryContext>
  );
};

const StyledContainer = styled(Container)`
  position: relative;
  margin-top: 10px;
`;

const StyledCard = styled(Card)`
  width: 18rem;
  margin-bottom: 20px;
`;
const StyledCardSubtitle = styled(Card.Subtitle)`
  margin-top: 3px !important;
  margin-bottom: 0px;
`;
const StyledCol = styled(Col)`
  height: 43rem;
  width: 33rem;
  overflow: scroll;
`;
const StyledImage = styled(Image)`
  padding-left: 5px;
  padding-right: 5px;
  padding-top: 5px;
`;

// const addProfilePhoto = () => {
//   const data = new FormData();
//   data.append('file', selectedFile);
//   axios.post(`${Config.apiBaseUrl}/profile-upload`, data).then(res => {
//     if (res.data) {
//       console.log(res.data);
//     }
//   });
// };
