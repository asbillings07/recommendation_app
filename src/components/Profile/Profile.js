import React, { useEffect, useState, useCallback } from 'react'
// import avatar from '../../images/imgholdr-image.png';
import ProfileRecommendation from './ProfileRecs'
import UserProfileInfo from './ProfileInfo'
import ProfilePhotoModal from './ProfilePhotoModal'
import { getUserById } from '../../Store/slices/userSlice'
import { useDispatch, useSelector } from 'react-redux'
import { Container, Row, Col, Card, Button } from 'react-bootstrap'
import { Image, CloudinaryContext } from 'cloudinary-react'
import styled from 'styled-components'

export const Profile = () => {
  const dispatch = useDispatch()
  const { user } = useSelector((state) => state.users)
  // const [user, setUser] = useState({})
  const [recommendation, setRecommendation] = useState([])
  // const [selectedFile, setSelectedFile] = useState('');
  const [photo, setPhoto] = useState('sample')
  const [showModal, setShowModal] = useState(false)

  const fetchData = useCallback(() => {
    dispatch(getUserById())
    // try {
    //   const user = await context.data.getUserById()
    //   console.log(user)
    //   if (user) {
    //     setUser(user)
    //     setRecommendation(user.Recommendations)
    //   }
    // } catch (err) {
    //   console.log(err.response)
    // }
  }, [dispatch])

  useEffect(() => {
    fetchData()
  }, [fetchData])

  const profilePhotos = [
    {
      photoName: 'Ninja Turtle Avatar',
      photoUrl: 'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401759/ninja-turtle.png'
    },
    {
      photoName: 'Penguin Avatar',
      photoUrl: 'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401755/penguin-avatar.png'
    },
    {
      photoName: 'Fox Avatar',
      photoUrl: 'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401753/fox-avatar.png'
    },
    {
      photoName: 'Ninja Avatar',
      photoUrl: 'https://res.cloudinary.com/hw8mbm47q/image/upload/v1568401753/ninja-avatar.png'
    }
  ]

  return (
    <CloudinaryContext cloudName='demo'>
      <StyledContainer>
        <Row>
          <Col sm={4}>
            <StyledCard aria-label='profile description'>
              <StyledImage
                publicId={user.imageId ? user.imageId : photo}
                width='286'
                crop='scale'
              />
              <Card.Body>
                <Card.Text>
                  This is your profile. You can reset your password, view and manage all of the
                  recommendations you have created and choose a profile Avatar
                </Card.Text>

                <Button onClick={() => setShowModal(true)} variant='primary'>
                  Update Profile Avatar
                </Button>
              </Card.Body>
            </StyledCard>
            <UserProfileInfo user={user} />
          </Col>

          <StyledCol sm={8}>
            <h1>Manage Recommendations</h1>
            <ProfileRecommendation
              recommendations={recommendation}
              // context={context}
              onRefresh={fetchData}
            />
          </StyledCol>
        </Row>
      </StyledContainer>
      <ProfilePhotoModal
        profilePhotos={profilePhotos}
        setModal={setShowModal}
        showModal={showModal}
        setPhoto={setPhoto}
        // context={context}
        refresh={fetchData}
      />
    </CloudinaryContext>
  )
}

const StyledContainer = styled(Container)`
  position: relative;
  margin-top: 10px;
`

const StyledCard = styled(Card)`
  width: 18rem;
  margin-bottom: 20px;
`
const StyledCol = styled(Col)`
  height: 43rem;
  width: 33rem;
  overflow: scroll;
`
const StyledImage = styled(Image)`
  padding-left: 5px;
  padding-right: 5px;
  padding-top: 5px;
`

// const addProfilePhoto = () => {
//   const data = new FormData();
//   data.append('file', selectedFile);
//   axios.post(`${Config.apiBaseUrl}/profile-upload`, data).then(res => {
//     if (res.data) {
//       console.log(res.data);
//     }
//   });
// };
