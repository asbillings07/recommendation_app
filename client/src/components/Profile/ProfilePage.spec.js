import React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  waitForElement,
} from '@testing-library/react';
import { Profile } from './Profile';
import { Provider } from '../../Context';
import ProfileInfo from './ProfileInfo';
import ProfileRecs from './ProfileRecs';
import ProfileEdit from './ProfileEdit';
import ProfileModal from './ProfileModal';

describe('Profile with all of its children render correctly', () => {
  afterEach(cleanup);

  const context = {
    token: '123456789',
  };
  const recommendations = [
    {
      id: 1,
      categoryId: 2,
      userid: 13,
      title: 'Test Recommendation 1',
      location: '125 main Street',
      description: 'test description 1',
    },
    {
      id: 2,
      categoryId: 2,
      userid: 13,
      title: 'New Test Recommendation 2',
      location: '124 main Street',
      description: 'test description 2',
    },
    {
      id: 3,
      categoryId: 2,
      userid: 13,
      title: 'Test Recommendation 3',
      location: '123 main Street',
      description: 'test description 3 ',
    },
  ];
  const user = {
    firstName: 'Aaron',
    lastName: 'Billings',
    email: 'test@test.com',
    confirmed: true,
  };

  test('<Profile/>', () => {
    const { queryByLabelText } = render(
      <Provider>
        <Profile />
      </Provider>
    );
    expect(queryByLabelText('profile description')).toBeTruthy();
  });

  test('<ProfileInfo/>', () => {
    const { queryByLabelText, getByText } = render(<ProfileInfo user={user} />);
    // ensure user profile information renders
    expect(queryByLabelText('Update Profile Information')).toBeTruthy();
    expect(
      queryByLabelText(`The first name is ${user.firstName}`)
    ).toBeTruthy();
    expect(queryByLabelText(`The last name is ${user.lastName}`)).toBeTruthy();
    expect(queryByLabelText(`The email is ${user.email}`)).toBeTruthy();
    expect(
      queryByLabelText(
        `Your email is ${user.confirmed ? 'CONFIRMED' : 'NOT CONFIRMED'}`
      )
    ).toBeTruthy();
    // ensure links are correct
    expect(getByText('Edit Profile Info').getAttribute('href')).toBe(
      '/profile/edit'
    );
    expect(getByText('Reset Password').getAttribute('href')).toBe(
      '/forgotpassword'
    );
    // ensure button fires correctly

    fireEvent.click(getByText('Edit Profile Info'));
    fireEvent.click(getByText('Reset Password'));
  });

  test('<ProfileRecs/>', () => {
    const { queryAllByLabelText, queryAllByText } = render(
      <ProfileRecs recommendations={recommendations} />
    );

    // ensure title, location and description render
    expect(queryAllByLabelText(`The title of the recommendation`)).toBeTruthy();
    expect(
      queryAllByLabelText(`The location of the recommendation `)
    ).toBeTruthy();
    expect(
      queryAllByLabelText(`The description of the recommendation`)
    ).toBeTruthy();

    // expect(getAllByText('Edit').getAttribute('href')).toBe(`/recs/update`);
    expect(queryAllByText('Delete')).toBeTruthy();
    //fireEvent.click(queryAllByText('Edit'));
  });

  test('<ProfileEdit/>', () => {
    const { container } = render(
      <Provider>
        <ProfileEdit context={context} user={user} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
