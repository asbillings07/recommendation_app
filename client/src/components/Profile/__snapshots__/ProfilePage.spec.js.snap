// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile with all of its children render correctly <ProfileEdit/> 1`] = `
<div
  class="mt-3 container"
>
  <div
    class="justify-content-md-center row"
  >
    <div
      class="col-lg-auto col-md col"
    >
      <h1
        class="mb-4"
      >
        Update Your Profile
      </h1>
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <input
            class="form-control"
            name="firstName"
            placeholder=""
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <input
            class="form-control"
            name="lastName"
            placeholder=""
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <input
            class="form-control"
            name="email"
            placeholder=""
            type="text"
            value=""
          />
        </div>
        <button
          class="mr-1 btn btn-primary"
          style="width: 13rem;"
          type="submit"
        >
          Update Profile
        </button>
        <button
          class="mr-1 btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </form>
    </div>
  </div>
</div>
`;
